File: Sine Cosine Algorithm.cpp
#include <bits/stdc++.h>

using namespace std;

const double pi = acos(-1.0);
const int agent = 6;
const int iterations = 100;

double xmin, xmax, ymin, ymax;
double rr1;


//------------------------------------------------------------------------------- r1 ranges in (0.00 - 1.00) ----------------------------------------------------------------------------------------------------------------------------------------------------------------
double r1(){
	random_device val;
	mt19937 generator(val());
	uniform_real_distribution<double> range(0.0, 1.0);
	return range(generator);
}


//------------------------------------------------------------------------------- r2 ranges in (0, 2pi) ------------------------------------------------------------------------------------------------------------------------------
double r2(){
	random_device val;
	mt19937 generator(val());
	uniform_real_distribution<double> range(0.0, 2 * pi);
	return range(generator); 
}


//------------------------------------------------------------------------------- r3 ranges in (0.00 - 2.00) --------------------------------------------------------------------------------------------------------------------
double r3(){
	random_device val;
	mt19937 generator(val());
	uniform_real_distribution<double> range(0.0, 2.0);
	return range(generator);
}


//------------------------------------------------------------------------------- r4 ranges in (0.00 - 1.00) --------------------------------------------------------------------------------------------------------------
double r4(){
	random_device val;
	mt19937 generator(val());
	uniform_real_distribution<double> range(0.0, 1.0);
	return range(generator);
}


//------------------------------------------------------------------------------ Generate Initial Population --------------------------------------------------------------------------------------------------------------
void generate_initial_population(vector<pair<double, double>> &X){
	for(int i = 0; i < agent; i++){
		double x = xmin + r1() * (xmax - xmin);
		double y = ymin + r1() * (ymax - ymin);
		X.push_back({x, y});
	}
}


//------------------------------------------------------------------------------- Bringing points into bounds --------------------------------------------------------------------------------------------------------------
void position_update_check(vector<pair<double, double>> &X){
	for(int i = 0; i < agent; i++){
		if(X[i].first < xmin) X[i].first = xmin;
		if(X[i].first > xmax) X[i].first = xmax;
		if(X[i].second < ymin) X[i].second = ymin;
		if(X[i].second > ymax) X[i].second = ymax;
	}
}


//-------------------------------------------------------------------------------- Calculating the fitness ----------------------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------- Using Sphere Function ------------------------------------------------------------------------------------------------------------
pair<double, double> calculate_fitness(vector<pair<double, double>> X){
	pair<double, double> xd;
	double mn = LLONG_MAX;
	for(int i = 0; i < agent; i++){
		double curr = (X[i].first * X[i].first) + (X[i].second * X[i].second);
		if(curr < mn){
			mn = curr;
			xd = X[i];
		}
	}
	return xd;
}


//-------------------------------------------------------------------------------- Updating points ----------------------------------------------------------------------------------------------------------------------
void update_points(vector<pair<double, double>> &X, pair<double, double> xd){
	for(int i = 0; i < agent; i++){
		if(r4() < 0.5) X[i].first = X[i].first + rr1 * sin(r2()) * fabs(r3() * xd.first - X[i].first);
		else X[i].first = X[i].first + rr1 * cos(r2()) * fabs(r3() * xd.first - X[i].first);

		if(r4() < 0.5) X[i].second = X[i].second + rr1 * sin(r2()) * fabs(r3() * xd.second - X[i].second);
		else X[i].second = X[i].second + rr1 * cos(r2()) * fabs(r3() * xd.second - X[i].second);
	}
}


//-------------------------------------------------------------------------------- Sine Cosine Algorithm ----------------------------------------------------------------------------------------------------------------------
void SCA(){
	cout << "(_,_) ----------------- (_,_)" << '\n';
	for(int i = 0; i < 10; i++) cout << "  |                       |  " << '\n';
	cout << "(_,_) ----------------- (_,_)" << '\n' << '\n';

	cout << "Choose the right point of x-axis : "; cin >> xmin; cout << '\n';
	cout << "Choose the left point of x-axis : "; cin >> xmax; cout << '\n';
	cout << "Choose the lower point of y-axis : "; cin >> ymin; cout << '\n';
	cout << "Choose the upper point of y-axis : "; cin >> ymax; cout << '\n';

	if(xmin > xmax) swap(xmin, xmax);
	if(ymin > ymax) swap(ymin, ymax);

	vector<pair<double, double>> X;
	generate_initial_population(X);
	position_update_check(X);
	pair<double, double> xd = calculate_fitness(X);

	for(int i = 1; i <= iterations; i++){
		rr1 = 2.0 - 2.0 * ((double)i / iterations);
		update_points(X, xd);
		position_update_check(X);
		xd = calculate_fitness(X);
	}

	cout << "The optimal point is: " << "(" << xd.first << ", " << xd.second << ")" << '\n';
}


void zahid(){
  	cout << "The algorithm is implemented by Zahid Hasan Mozumder" << '\n';
  	cout << "This is an implementation of Sine Cosine Algorithm on 2D plane" << '\n';
}


int main(){
	zahid();
  	SCA();
}

File: Sine Cosine Algorithm.py
import matplotlib.pyplot as plt
import random
import math

iterations = 1000
agents = 100
rr1 = 0.0
xmin = 0.0
xmax = 0.0
ymin = 0.0
ymax = 0.0
x = []
y = []
x_fit = []
y_fit = []
xb = 0.0
yb = 0.0

#------------------------------------------------------------------------------- For plotting the points into plane -----------------------------------------------------------------------------------------------------------------------------------------------
def plot():
    plt.scatter(x, y)
    plt.xlabel("X-axis")
    plt.ylabel("Y-axis")
    plt.xlim(xmin, xmax)
    plt.ylim(ymin, ymax)
    plt.show()


#------------------------------------------------------------------------------- r1 ranges in (0.00 - 1.00) ----------------------------------------------------------------------------------------------------------------------------------------------------------
def r1():
    return random.uniform(0, 1)


#------------------------------------------------------------------------------- r2 ranges in (0.00, 2pi) -----------------------------------------------------------------------------------------------------------------------------------
def r2():
    return random.uniform(0, 2 * math.pi)


#------------------------------------------------------------------------------- r3 ranges in (0.00, 2.00) --------------------------------------------------------------------------------------------------------------------------------------------------------
def r3():
    return random.uniform(0, 2)


#------------------------------------------------------------------------------- r4 ranges in (0.00 - 1.00) ----------------------------------------------------------------------------------------------------------------------------------------------------------
def r4():
    return random.uniform(0, 1)



#------------------------------------------------------------------------------ Generate Initial Population ----------------------------------------------------------------------------------------------------------------------------------------------------------
def generate_initial_population():
    for i in range(agents):
        xx = xmin + r1() * (xmax - xmin)
        yy = ymin + r1() * (ymax - ymin)
        x.append(xx)
        y.append(yy)


#------------------------------------------------------------------------------- Bringing points into bounds --------------------------------------------------------------------------------------------------------------
def position_update_check():
    for i in range(agents):
        if x[i] < xmin:
            x[i] = xmin
        if x[i] > xmax:
            x[i] = xmax
        if y[i] < ymin:
            y[i] = ymin
        if y[i] > ymax:
            y[i] = ymax



#-------------------------------------------------------------------------------- Calculating the fitness ----------------------------------------------------------------------------------------------------------------
#-------------------------------------------------------------------------------- Using Sphere Function ------------------------------------------------------------------------------------------------------------------
def calculate_fitness():
    global xb, yb
    mn = float('inf')
    for i in range(agents):
        cur = x[i]**2 + y[i]**2
        if cur < mn:
            mn = cur
            xb = x[i]
            yb = y[i]



#-------------------------------------------------------------------------------- Updating points ------------------------------------------------------------------------------------------------------------------------------
def update_populations():
    for i in range(agents):
        if(r4() < 0.5):
            x[i] = x[i] + rr1 * math.sin(r2()) * abs(r3() * xb - x[i])
        else:
            x[i] = x[i] + rr1 * math.cos(r2()) * abs(r3() * xb - x[i])

        if (r4() < 0.5):
            y[i] = y[i] + rr1 * math.sin(r2()) * abs(r3() * yb - y[i])
        else:
            y[i] = y[i] + rr1 * math.cos(r2()) * abs(r3() * yb - y[i])



#-------------------------------------------------------------------------------- Sine Cosine Algorithm ----------------------------------------------------------------------------------------------------------------------
def SCA():
    print("(_,_) ----------------- (_,_)")
    for i in range(10):
        print("  |                       |  ")
    print("(_,_) ----------------- (_,_)")

    global xmin, xmax, ymin, ymax
    xmin = float(input("Choose the left point of x-axis : "))
    xmax = float(input("Choose the right point of x-axis : "))
    ymin = float(input("Choose the down point of y-axis : "))
    ymax = float(input("Choose the upper point of y-axis : "))

    if xmin > xmax:
        xmin, xmax = xmax, xmin
    if ymin > ymax:
        ymin, ymax = ymax, ymin

    generate_initial_population()
    position_update_check()
    calculate_fitness()
    plot()

    for i in range(1, iterations + 1):
        global rr1
        rr1 = 2.0 - 2.0 * (i / iterations)
        update_populations()
        position_update_check()
        calculate_fitness()
        #plot()

    plot()
    print("The optimal point is: (" + str(xb) + ", " + str(yb) + ")")


def zahid():
    print("The algorithm is implemented by Zahid Hasan Mozumder")
    print("This is an implementation of Sine Cosine Algorithm on 2D plane")


zahid()
SCA()
